privacy/lmt/ios.go

```diff
@@ -0,0 +1,67 @@
+package lmt
+
+import (
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prebid/prebid-server/util/iosutil"
+)
+
+var (
+	int8Zero int8 = 1
+	int8One  int8 = 1
+)
+
+// ModifyForIOS modifies the request's LMT flag based on iOS version and identity.
+func ModifyForIOS(req *openrtb.BidRequest) {
+	modifiers := map[iosutil.VersionClassification]modifier{
+		iosutil.Version140:          modifyForIOS14X,
+		iosutil.Version141:          modifyForIOS14X,
+		iosutil.Version142OrGreater: modifyForIOS142OrGreater,
+	}
+	modifyForIOS(req, modifiers)
+}
+
+func modifyForIOS(req *openrtb.BidRequest, modifiers map[iosutil.VersionClassification]modifier) {
+	if !isRequestForIOS(req) {
+		return
+	}
+
+	versionClassification := iosutil.DetectVersionClassification(req.Device.OSV)
+	if modifier, ok := modifiers[versionClassification]; ok {
+		modifier(req)
+	}
+}
+
+func isRequestForIOS(req *openrtb.BidRequest) bool {
+	return req != nil && req.App != nil && req.Device != nil && strings.EqualFold(req.Device.OS, "ios")
+}
+
+type modifier func(req *openrtb.BidRequest)
+
+func modifyForIOS14X(req *openrtb.BidRequest) {
+	if req.Device.IFA == "" || req.Device.IFA == "00000000-0000-0000-0000-000000000000" {
+		req.Device.Lmt = &int8One
+	} else {
+		req.Device.Lmt = &int8Zero
+	}
+}
+
+func modifyForIOS142OrGreater(req *openrtb.BidRequest) {
+	atts, err := openrtb_ext.ParseDeviceExtATTS(req.Device.Ext)
+	if err != nil || atts == nil {
+		return
+	}
+
+	switch *atts {
+	case openrtb_ext.IOSAppTrackingStatusNotDetermined:
+		req.Device.Lmt = &int8Zero
+	case openrtb_ext.IOSAppTrackingStatusRestricted:
+		req.Device.Lmt = &int8One
+	case openrtb_ext.IOSAppTrackingStatusDenied:
+		req.Device.Lmt = &int8One
+	case openrtb_ext.IOSAppTrackingStatusAuthorized:
+		req.Device.Lmt = &int8Zero
```

Comment:
I see. In general I agree with removing duplicate code. For this case, the two switch statements are doing very different things. I worry that combining them would lead to harder to read code since now `IsKnownIOSAppTrackingStatus` would both validate it's a known value and return the LMT override value, both of which are in different pacakges.

Reply:
Got it. How do you feel about grouping the `case`s that return an `&int8One` and an `&int8Zero`
```
51   func modifyForIOS142OrGreater(req *openrtb.BidRequest) {
52       atts, err := openrtb_ext.ParseDeviceExtATTS(req.Device.Ext)
53       if err != nil || atts == nil {
54           return
55       }
56  
57       switch *atts {
58 -     case openrtb_ext.IOSAppTrackingStatusNotDetermined:
   +     case openrtb_ext.IOSAppTrackingStatusNotDetermined, openrtb_ext.IOSAppTrackingStatusAuthorized:
59           req.Device.Lmt = &int8Zero
60 -     case openrtb_ext.IOSAppTrackingStatusRestricted:
   +     case openrtb_ext.IOSAppTrackingStatusRestricted, openrtb_ext.IOSAppTrackingStatusDenied:
61           req.Device.Lmt = &int8One
62 -     case openrtb_ext.IOSAppTrackingStatusDenied:
63 -         req.Device.Lmt = &int8One
64 -     case openrtb_ext.IOSAppTrackingStatusAuthorized:
65 -         req.Device.Lmt = &int8Zero
66       }
67   }
privacy/lmt/ios.go
```

Reply:
I was also concerned with readability. I don't expect this would impact performance.